app/scripts/directives/istagSelect.js

```diff
@@ -21,31 +25,102 @@ angular.module("openshiftConsole")
+                $scope.isByNamespace[ns][$scope.istag.imageStream].status.tags.push({tag: $scope.istag.tagObject.tag});
+              }
+            });
+          }
+        };
+
         DataService.list("projects", {}, function(projectData) {
-          var sortedNamespaces = _.keys(projectData.by('metadata.name')).sort();
+          $scope.namespaces = _.keys(projectData.by('metadata.name')).sort();
 
-          // Use _.uniq to avoid adding "openshift" twice if the user is a
-          // member of the openshift namespace.
-          $scope.namespaces = _.uniq(['openshift'].concat(sortedNamespaces));
+          if ($scope.includeOpenshift) {
+            // Use _.uniq to avoid adding "openshift" twice if the user is a
+            // member of the openshift namespace.
+            $scope.namespaces = _.uniq(['openshift'].concat($scope.namespaces));
+          }
 
           // Fetch image streams when a new namespace is picked.
           $scope.$watch('istag.namespace', function(namespace) {
+
             if (!namespace || $scope.isByNamespace[namespace]) {
               // We already have the data (or nothing selected).
               return;
             }
 
-            DataService.list('imagestreams', { namespace: namespace }, function(isData) {
-              $scope.isByNamespace[namespace] = isData.by('metadata.name');
-              $scope.isNamesByNamespace[namespace] = _.keys($scope.isByNamespace[namespace]).sort();
-            });
+            if ($scope.prepopulate) {
+              prepopulate(namespace);
+              $scope.prepopulate = false;
+            } else {
+              DataService.list('imagestreams', { namespace: namespace }, function(isData) {
+                $scope.isByNamespace[namespace] = isData.by('metadata.name');
+
+                _.each(_.keys($scope.isByNamespace[namespace]), function(imageStream) {
+                  if (!$scope.isByNamespace[namespace][imageStream].status.tags) {
+                    $scope.isByNamespace[namespace][imageStream].status = { tags: []};
+                  }
+                });
+
+                $scope.isNamesByNamespace[namespace] = _.keys($scope.isByNamespace[namespace]).sort();
+              });
+            }
+
+            $scope.getTags = function(search) {
+              if ($scope.customTagValue && search && !_.find($scope.isByNamespace[$scope.istag.namespace][$scope.istag.imageStream].status.tags, {tag: search})) {
+                _.remove($scope.isByNamespace[$scope.istag.namespace][$scope.istag.imageStream].status.tags, function(tag) {
+                  return !tag.items;
+                });
+                $scope.isByNamespace[$scope.istag.namespace][$scope.istag.imageStream].status.tags.unshift({tag: search});
+              }
+            };
+
+            $scope.groupTags = function(tagObject) {
+              if ($scope.customTagValue) {
+                if (tagObject.items) {
+                return 'AVAILABLE TAGS';
```

Comment:
I feel like we should use `text-transform: capitalize;` style on the group header instead of returning all caps here if it's what we want.

Indentation is off here.